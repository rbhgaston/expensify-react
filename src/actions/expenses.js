// import {v4 as uuidv4} from 'uuid';
import db from '../firebase/firebase'

// ADD
export const  addExpense = (expense) => ({
    type: 'ADD',
    expense
})

export const startAddExpense = ({description= '', note= '', createdAt= 0, amount= 0} = {}) =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        const expense = {
            description,
            amount,
            note,
            createdAt
        }
        return db.ref(`users/${uid}/expenses`).push(expense)
            .then(ref => dispatch(addExpense({
                id: ref.key,
                ...expense
            })))
    }

// EDIT
export const editExpense = (id, updates) => ({
    type: 'EDIT',
    id,
    updates
})

export const startEditExpense= (id, updates) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).update(updates)
            .then(() => dispatch(editExpense(id, updates)))
    }



// REMOVE
export const removeExpense = (id) => ({type: 'REMOVE', id})

export const startRemoveExpense = (id) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).remove()
            .then(() => dispatch(removeExpense(id)))
        }

// SET
export const setExpenses = (expenses) => ({
    type: 'SET',
    expenses
})

export const startSetExpenses = () =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses`).once('value').then(snapshot => {
            const expensesArray = []
            snapshot.forEach(expense => {expensesArray.push({id: expense.key.toString(), ...expense.val()})})
            dispatch(setExpenses(expensesArray))
        })
    }