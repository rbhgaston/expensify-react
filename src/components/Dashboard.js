import React from "react";
import ExpensesList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

export default function DashBoard() {
    return (
        <div>
            <ExpenseListFilters />
            <ExpensesList />
        </div>
    )
}