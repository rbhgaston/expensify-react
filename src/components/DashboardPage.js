import React from "react";
import ExpensesList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseListSummary from './ExpenseListSummary';

export default function DashboardPage() {
    return (
        <div>
            <ExpenseListSummary />
            <ExpenseListFilters />
            <ExpensesList />
        </div>
    )
}