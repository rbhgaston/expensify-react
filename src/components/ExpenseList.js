import React from 'react'
import {connect} from 'react-redux'
import ExpenseItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => {
    return (
        <div>
            {
                !props.expenses.length ?
                    <p>No expenses</p>
                    : props.expenses.map(expense =>
                        <ExpenseItem key={expense.id} {...expense}/>
                    )
            }

        </div>
    )
}
const mapStateToProps = (state) => ({expenses: selectExpenses(state.expenses, state.filters)})
const ConnectedExpensesList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpensesList