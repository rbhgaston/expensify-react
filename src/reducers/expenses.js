const expensesReducerDefaultState = []
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD':
            return [
                ...state,
                action.expense
            ]
        case 'EDIT':
            return state.map(expense => expense.id === action.id ? {...expense, ...action.updates} : expense )
        case 'REMOVE':
            return state.filter(expense => expense.id !== action.id)
        default:
            return state
    }
}

export default expensesReducer