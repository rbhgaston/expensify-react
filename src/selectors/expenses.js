import moment from 'moment';

export const getFilteredExpenses =  (expenses, {text, sort, startDate, endDate}) => {
    return expenses.filter(expense => {
        const isGreaterThanStart = startDate ?  moment(expense.createdAt).isSameOrAfter(startDate, 'day') : true
        const isLessThanEnd = endDate ? moment(expense.createdAt).isSameOrBefore(endDate, 'day') : true

        return expense.description.includes(text) && isGreaterThanStart && isLessThanEnd
        })
        .sort((a, b) => sort === 'amount' ? a.amount - b.amount : a.createdAt - b.createdAt)
}

export const getTotalAmount = (expenses = []) => expenses.reduce((total, expense) => total + expense.amount, 0)