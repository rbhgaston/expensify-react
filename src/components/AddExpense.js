import React from 'react';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';
import {connect} from 'react-redux'

export function AddExpense(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.addExpense(expense)
                    props.history.push('/')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

// if you pass mapDispatchToProps, you lose access to dispatch in props
export default connect(undefined, mapDispatchToProps)(AddExpense)

