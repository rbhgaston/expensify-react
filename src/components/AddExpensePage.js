import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {startAddExpense, addExpense} from '../actions/expenses';

export function AddExpensePage(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.startAddExpense(expense)
                    props.history.push('/')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

// if you pass mapDispatchToProps, you lose access to dispatch in props
export default connect(undefined, mapDispatchToProps)(AddExpensePage)

