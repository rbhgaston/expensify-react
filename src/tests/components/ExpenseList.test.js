import React from 'react';
import {ExpenseList} from '../../components/ExpenseList';
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'


test('render ExpenseList component with data provided', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test('render ExpenseList component with an empty array', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('p').text()).toBe('No expenses')
})