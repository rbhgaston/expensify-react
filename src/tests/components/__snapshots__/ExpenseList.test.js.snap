// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseList component with an empty array 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`render ExpenseList component with data provided 1`] = `
<div>
  <ExpenseListItem
    amount={100}
    createdAt={777600000}
    description="test0"
    id={0}
    key="0"
  />
  <ExpenseListItem
    amount={300}
    createdAt={691200000}
    description="test1"
    id={1}
    key="1"
  />
  <ExpenseListItem
    amount={200}
    createdAt={604800000}
    description="test2"
    id={2}
    key="2"
  />
</div>
`;
